src/fenics_constitutive/solver.py

```diff
@@ -0,0 +1,318 @@
+from __future__ import annotations
+
+import basix
+import dolfinx as df
+import numpy as np
+import ufl
+from petsc4py import PETSc
+
+from .interfaces import IncrSmallStrainModel
+from .maps import SubSpaceMap, build_subspace_map
+from .stress_strain import ufl_mandel_strain
+
+
+def build_history(
+    law: IncrSmallStrainModel, mesh: df.mesh.Mesh, q_degree: int
+) -> df.fem.Function | dict[str, df.fem.Function] | None:
+    """Build the history space and function(s) for the given law.
+
+    Args:
+        law: The constitutive law.
+        mesh: Either the full mesh for a homogenous domain or the submesh.
+        q_degree: The quadrature degree.
+
+    Returns:
+        The history function(s) for the given law.
+
+    """
+    match law.history_dim:
+        case int():
```

Comment:
`case int()` means that we have a single history variable that is a tensor of rank 2 (such as back stress for plasticity with kinematic hardening)?

Also, in case we have more than one history variable, we always need to use a `dict` is that correct? If so, I think 
that is reasonable.

However, I think the case where we have a single history variable that is a fouth order tensor is currently missing, i.e. `case tuple():`. Not sure if there exists a material model where this may be relevant.

Reply:
My thought was more along the line that we use dictionaries for tensors, even if there is only one. The `int()` option is for wrapping an Abaqus UMAT where the history is just a vector and only the UMAT knows what element of the vector means what.

But we could also say that the binding of a UMAT could have the history as a dictionary with one entry like `{"umat_history": ...}`. This would also eliminate one branch in some `match` or `if` statements.

Reply:
> My thought was more along the line that we use dictionaries for tensors, even if there is only one. The int() option is for wrapping an Abaqus UMAT where the history is just a vector and only the UMAT knows what element of the vector means what.

I see. That means if I have a scalar variable, a rank 1 tensor and a rank 2 tensor (in total 3 variables) as history variables, then I would simply put `history_dim: int = 10`.

Either is fine with me, and once you add the docstrings there probably will be less room for confusion.